### ISSUE TITLE
FAQ: Regularizing the noise covariance matrix

### ISSUE THREAD

**Issue by christianbrodbeck** [mne-tools/mne-python#1637]
re: https://github.com/mne-tools/mne-python/issues/1495#issuecomment-57097141 I have a first question (I will start a FAQ with it as @agramfort suggested): How should I regularize my noise covariance matrix?

The [manual](http://martinos.org/mne/stable/manual/mne.html#regularization-of-the-noise-covariance-matrix) says "Since finite amount of data is usually available... " does that mean after having a certain amount of data regularization is no longer required?

What are recommended values for regularization? I noticed that the [example](http://martinos.org/mne/stable/auto_examples/inverse/plot_make_inverse_operator.html) does not use the function's default values. What do these values depend on?

If regularization is something that should always be done, should the function be more prominent? E.g., a `Covariance` method? Or even an argument of `make_inverse_operator()`?

**Comment by teonlamont** [comment]
hey. is there a recommendation for the number of epochs one should use to calculate the covariance matrix? I have an experiment where there were around 1000 trials and in the past I use baseline of all of the trials. I've tried to do this with the 'auto' parameter and this process did not complete (I waited an hr and it was still iterating through the Factor Analysis).

**Comment by christianbrodbeck** [comment]
I think Dennis mentioned something when calculating with traditional method in my FAQ github thread (1000 should be enough IIRC). I don’t know about the auto method. Would probably be a good issue to ping the FAW thread with...

**Comment by agramfort** [comment]
if you have a lot of epochs FA will take too much time. Just use
method=["shrunk","empirical"]. Let me know how it goes.

**Comment by teonlamont** [comment]
i let it run overnight and overwhelmingly, shrunk was chosen, with the occasional minor victories for factor analysis.

**Comment by Eric89GXL** [comment]
I think that we should change `method='auto'` to triage based on the number of epochs; if > 100 (or whatever @agramfort and @dengemann recommend) it should be equivalent to `method=["shrunk", "empirical"]` instead of testing the fuller set. `auto` isn't very useful if it locks up your computer for 12h to make insignificant modifications to the matrix.

**Comment by agramfort** [comment]
+1

**Comment by dengemann** [comment]
It depends on more things, especially the effective sample size which is determined by the low pass filter. Also you normally won't see FA winning over shrunk on single sensor data, especially after rank reduction.

**Comment by dengemann** [comment]
I would e.g. exclude fa by default if data are SSSed.

**Comment by Eric89GXL** [comment]
I'm not sure what you mean by "single sensor data"...?

Triaging based on SSS is okay with me if that really helps, but I still think it would be nice if we could triage based on the number of epochs, plus whatever other factors matter. It doesn't have to be perfect, but we should definitely prevent freezing someone's computer for so long (I had a similar disappointing experience). @dengemann would you be satisfied if we used a rule of thumb based on number of epochs scaled by the relative width of the pass-band? Do you have a sense if this would be the right scaling?

**Comment by dengemann** [comment]
@Eric89GXL by single sensor I mean single sensor type, e.g. mag or grad or eeg only. FA really wins when signals are more complex.

**Comment by Eric89GXL** [comment]
Is that true even when the number of epochs are high?

**Comment by dengemann** [comment]
> Is that true even when the number of epochs are high?

FA tends to benefit from having more samples available. It will not converge if data are SSSed or heavily projected, so the procedure will climb up all possible ranks, adding more computaiton time at each iteration.

**Comment by Eric89GXL** [comment]
I set up a script to test covariance calculation using 20, 40, or 80 epochs with `method='auto', return_residuals=True` with a few subjects. I am getting `-np.inf` for log likelihoods for ~68% of the estimators. Is this expected? I was hoping to use it to assess quality, but with that many -np.infs it's pretty useless. (In most cases, only one estimator returns a non-inf log likelihood.) I'm using a simple command to generate the covariances:
```
                cov = mne.compute_covariance(epochs_all[:n_epochs],
                                             method='auto',
                                             return_estimators=True)
```
Is there some parameter I need to tweak in order to get meaningful log likelihoods? Do I need more folds in my cross-validation or something?

**Comment by Eric89GXL** [comment]
A bit more info -- I set up some tests using three subjects (to start) with the following parameters:

1. Three data conditions: real empty-room data, a `RawArray` instance where the data is independent Gaussian noise, and a `RawArray` with rank-deficient gaussian noise (by applying a projection operation).

2. Low-passing at 2 Hz and not low-passing at all.

3. Computing covariance using 20/40/80 epochs.

And I get the ~68% useless log likelihoods, pretty evenly spread across these 3*2*2=12 conditions -- even for the sanest condition (real data, 80 epochs, no low-pass) I get `-np.inf` for all estimators except  "empirical". Any ideas?

**Comment by dengemann** [comment]
Mhm, normally emprical tends to produce infs. Why do you use a low pass of 2 hz?

**Comment by dengemann** [comment]
Is your data sssed? It's not so uncommon to see infs with few trials, most of them come from the determinant in the log likelihood function in sklearn. Could you share a script, Eric?

**Comment by Eric89GXL** [comment]
No, they are not SSS'ed. Don't worry about the 2 Hz, the problem persists in both cases. But in case you're curious, I was going to look at what you said about how the filtering matters. I was going to look at what happens when systematically introducing (temporal) correlations between samples to see if there was a relationship between some "effective number of samples" and the best method, or at least to see if there's some number of effective samples where we can not run factor_analysis. Even with my simple test set it took a loooong time to compute, which makes me wary of using `method='auto'` currently.

**Comment by agramfort** [comment]
getting inf with shrunk is not really understandable. By construction the
estimated cov is invertible.

what version of sklearn do you use?

do you see the pb with sample or somato?

**Comment by Eric89GXL** [comment]
I use some dev version from a few months ago. I'll try using the sample dataset to see if I can reproduce the problem there, and also try updating sklearn.

**Comment by agramfort** [comment]
keep us posted.

**Comment by dengemann** [comment]
> Even with my simple test set it took a loooong time to compute, which makes me wary of using method='auto' currently.

That's really weird. If you don't feed it with wrong data FA completes after 5-40 seconds.

**Comment by dengemann** [comment]
@Eric89GXL very dumb, but I figured that below 40-30 Hz of lowpass filtering FA has too few samples. Above it can benefit from it.

**Comment by Eric89GXL** [comment]
Using empty-room data from a single subject, I am getting 71% of the log likelihoods as -np.inf. This is for half of the tests using real data, the other half with the `raw._data` replaced with independent gaussian noise. This is with sklearn 0.16-dev from a couple of months ago.

When I switch to using the `sample` dataset, I get 58% of the values as -np.inf.

I updated to the latest dev sklearn and got the same result. Feel free to try on your machine and see if it still happens:

https://gist.github.com/Eric89GXL/83f6184d9a7123cf9441

Maybe I'm doing something stupid?

**Comment by Eric89GXL** [comment]
I also just ran that file on my Windows machine running Py3k Anaconda w/all stock numpy/scipy/sklearn, and got 54% -np.inf.

**Comment by dengemann** [comment]
https://www.dropbox.com/s/5qmqiyasqqlh5ci/Screenshot%202015-02-15%2016.03.03.png?dl=0

Eric the data you use don't have any spatial structure. I'm not suprised about the infs here. It's not a relevant data set. Otherwise infs in empirical and diagonal_fixed are common if the input data is not sufficient. The log det will produce infs as the matrix is not invertible on some folds.

**Comment by Eric89GXL** [comment]
So you're saying the method is expected to fail on independent noise?

**Comment by Eric89GXL** [comment]
BTW I don't know why there is no structure, esp. for the sample dataset. It doesn't really make any sense...?

**Comment by Eric89GXL** [comment]
Naively I would have assumed that It seems like the simplest / easiest case to detect would be when the channels are all independent and full rank, so I'm really confused by this.

**Comment by agramfort** [comment]
I agree with Eric. There is something to understand here. I am taking a
look now.

**Comment by dengemann** [comment]
@Eric89GXL you're right, it should work. It turns out that your input data is not scaled. For numerical reasons we rescale the input data, assuming them to be in M/EEG measurement units. Scaling your data should do the trick and give you reasonable results.

**Comment by Eric89GXL** [comment]
I'm always worried with assumptions / caveats like this -- needing a particular set of scalings in your data for things to work correctly just seems like begging for a failure down the line.

IIRC it fails not just on the noise but also on the (real) sample dataset, though, too...?

**Comment by Eric89GXL** [comment]
Although I guess we might have to assume for numerical reasons that channels are scaled within a couple orders of magnitude of their expected range, so I suppose that's okay. Did it work correctly for the actual (sample) dataset case for you, though? Those data should be scaled realistically. I'll try implementing reasonable scale-factors on my gaussian noise, too, to see how it works.

**Comment by dengemann** [comment]
> needing a particular set of scalings in your data for things to work correctly just seems like begging for a failure down the line.

well but many sklearn algos would not work on the femto tesla scale. I don't see how rescaling could be avoided. It's the same thing with our rank estimators.
 
> IIRC it fails not just on the noise but also on the (real) sample dataset, though, too...?

IIRC it does not fail on the real data set (I tested this extensively ...) 
'empirical' tends to produce infs, 'diagonal_fixed' as well if the regularisation parameter is not good.

**Comment by Eric89GXL** [comment]
Ahh okay, I'll modify and test

**Comment by agramfort** [comment]
Eric keep us posted. You could also test when only 1 channel type is
present. In this case it should really work 99.999% of the times.

**Comment by Eric89GXL** [comment]
I added scalings to the random independent noise case, and get the following:
```
inf: 0.416666666667% for real
inf: 0.5% for noise
```
So it still doesn't make sense to me why this isn't working. @dengemann you said you didn't have problems with the sample dataset, can you try running the latest gist and see if you get the same percentages? I'm not sure if you actually tested the gist I linked to or did your own testing:

https://gist.github.com/Eric89GXL/83f6184d9a7123cf9441

**Comment by dengemann** [comment]
@Eric89GXL 50% infs is more or less expected. empircal and diagonal_fixed.

**Comment by Eric89GXL** [comment]
I ran a bunch of experimental, non-SSS'ed data (15 subjects) and got this result:
```
inf:  45.1041666667% for real:
    (97.0833333333% for diagonal_fixed)
    (78.75% for empirical)
    (2.08333333333% for factor_analysis)
    (2.5% for shrunk)
```
And these are my log likelihoods, averaged across subjects:
![real](https://cloud.githubusercontent.com/assets/2365790/6240054/d77ad838-b6be-11e4-9da9-70df73529a1c.png)
Looks like factor analysis wins on average across the board for me. I'm a bit surprised that the scores are about the same regardless of the lowpass, too, but maybe that's the 1/f spectrum falloff coming into play.

**Comment by agramfort** [comment]
still something to fix? shall we dig into the scoring thing?

We should have a bit of time this we

**Comment by Eric89GXL** [comment]
I still don't like the -np.inf scores. This to me means that the likelihood
is zero, which doesn't really make sense to me. Perhaps more imprudent, it
makes the measure useless for evaluating the quality of empirical and
diagonal fixed. I also occasionally got them for shrunk and factor
analysis, which isn't good. I can track down those examples if its helpful.

**Comment by agramfort** [comment]
have a look if you can. Otherwise we'll look into it this we with Denis

**Comment by Eric89GXL** [comment]
I probably won't have time to look into the mne-python covariance calculation code for at least a couple of weeks unfortunately. I can get you the files if that helps, though, but you guys must have seen this on your files, too?
